homeassistant/components/aemet/weather_update_coordinator.py

```diff
@@ -0,0 +1,652 @@
+    ATTR_DATA,
+)
+from aemet_opendata.helpers import (
+    get_forecast_day_value,
+    get_forecast_hour_value,
+    get_forecast_interval_value,
+)
+import async_timeout
+
+from homeassistant.components.weather import (
+    ATTR_FORECAST_CONDITION,
+    ATTR_FORECAST_PRECIPITATION,
+    ATTR_FORECAST_PRECIPITATION_PROBABILITY,
+    ATTR_FORECAST_TEMP,
+    ATTR_FORECAST_TEMP_LOW,
+    ATTR_FORECAST_TIME,
+    ATTR_FORECAST_WIND_BEARING,
+    ATTR_FORECAST_WIND_SPEED,
+)
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    ATTR_API_CONDITION,
+    ATTR_API_FORECAST,
+    ATTR_API_HUMIDITY,
+    ATTR_API_PRESSURE,
+    ATTR_API_RAIN,
+    ATTR_API_RAIN_PROB,
+    ATTR_API_SNOW,
+    ATTR_API_SNOW_PROB,
+    ATTR_API_STATION_ID,
+    ATTR_API_STATION_NAME,
+    ATTR_API_STATION_TIMESTAMP,
+    ATTR_API_STORM_PROB,
+    ATTR_API_TEMPERATURE,
+    ATTR_API_TEMPERATURE_FEELING,
+    ATTR_API_TOWN_ID,
+    ATTR_API_TOWN_NAME,
+    ATTR_API_TOWN_TIMESTAMP,
+    ATTR_API_WIND_BEARING,
+    ATTR_API_WIND_MAX_SPEED,
+    ATTR_API_WIND_SPEED,
+    CONDITIONS_MAP,
+    DOMAIN,
+    FORECAST_MODE_DAILY,
+    FORECAST_MODE_HOURLY,
+    WIND_BEARING_MAP,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+WEATHER_UPDATE_INTERVAL = timedelta(minutes=10)
+
+
+def format_condition(condition: str) -> str:
+    """Return condition from dict CONDITIONS_MAP."""
+    for key, value in CONDITIONS_MAP.items():
+        if condition in value:
+            return key
+    _LOGGER.error('condition "%s" not found in CONDITIONS_MAP', condition)
+    return condition
+
+
+def format_float(value) -> float:
+    """Try converting string to float."""
+    try:
+        return float(value)
+    except ValueError:
+        return None
+
+
+def format_int(value) -> int:
+    """Try converting string to int."""
+    try:
+        return int(value)
+    except ValueError:
+        return None
+
+
+class TownNotFound(UpdateFailed):
+    """Raised when town is not found."""
+
+
+class WeatherUpdateCoordinator(DataUpdateCoordinator):
+    """Weather data update coordinator."""
+
+    def __init__(self, aemet, latitude, longitude, forecast_mode, hass):
+        """Initialize coordinator."""
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=WEATHER_UPDATE_INTERVAL
+        )
+
+        self._aemet = aemet
+        self._station = None
+        self._town = None
+        self._latitude = latitude
+        self._longitude = longitude
+        self._forecast_mode = forecast_mode
+        self._data = {
+            "daily": None,
+            "hourly": None,
+            "station": None,
+        }
+
+    async def _async_update_data(self):
+        data = {}
+        with async_timeout.timeout(120):
+            weather_response = await self._get_aemet_weather()
+            data = self._convert_weather_response(weather_response)
```

Comment:
Please only wrap the line that can raise in the try... except block.

Reply:
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/weather_update_coordinator.py#L73